### PR TITLE
Add file system library

### PR DESCRIPTION
**Pull request by nilsdeppe**
Adds C++ wrappings around the POSIX file system calls to provide an easier to use interface.

### CODE REVIEW

File: src/Utilities/FileSystem.hpp
Comment (wthrowe) [inline src/Utilities/FileSystem.hpp]: "Given a path ... returns the path"?  I haven't looked at the source yet, but I assume this isn't an identity function.

File: src/Utilities/FileSystem.hpp
Comment (wthrowe) [inline src/Utilities/FileSystem.hpp]: Should note whether this resolves symlinks.

File: src/Utilities/FileSystem.hpp
Comment (wthrowe) [inline src/Utilities/FileSystem.hpp]: How about "including any parents that don't exist"?

Comment (wthrowe) [reply src/Utilities/FileSystem.hpp]: What happens if the directory already exists?

File: src/Utilities/FileSystem.hpp
Comment (wthrowe) [inline src/Utilities/FileSystem.hpp]: FileSystem

File: src/Utilities/FileSystem.hpp
Comment (wthrowe) [inline src/Utilities/FileSystem.hpp]: Do we really need to document that arguments must be of the type in the function signature?

File: src/Utilities/FileSystem.hpp
Comment (wthrowe) [inline src/Utilities/FileSystem.hpp]: And if `recursive` is false?

File: src/Utilities/FileSystem.hpp
Comment (wthrowe) [inline src/Utilities/FileSystem.hpp]: `num_tries` should not be signed.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: This `and` should probably be an `or`, although I'm still not sure what this function is doing.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: Why is this line excluded?

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: This is going to return something the caller is probably not prepared to handle if `file_path` ends in a `/`.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: Memory leak: `name` is allocated with `malloc` and never freed.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: Control can get past here, and if it does we segfault.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: Crash on `dir == "/"`.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: `check_if_file_exists` and `is_file` seem to have different ideas of what a file is.  Also, since this doesn't agree with the meaning of the unixish term "file", what condition is checked should be documented very clearly.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp:184]: Memory leak.

File: tests/Unit/Utilities/Test_FileSystem.cpp
Comment (wthrowe) [inline tests/Unit/Utilities/Test_FileSystem.cpp]: Check that it worked?

**Review by wthrowe — CHANGES_REQUESTED** [review]
I'm leaving most of `create_directory` for later.  We might be able to use some of the path manipulation functions to simplify it, depending on how those turn out.

File: src/Utilities/FileSystem.cpp
Comment (fmahebert) [inline src/Utilities/FileSystem.cpp]: Is there value in this historical note? Breaking on success is not surprising behavior.

I suggest deleting lines 116 ("We used to...") and 117 (whitespace).

File: tests/Unit/Utilities/Test_FileSystem.cpp
Comment (fmahebert) [inline tests/Unit/Utilities/Test_FileSystem.cpp]: I'm not sure what the intended behavior is for `get_path("/test/path/to/dir")` without a final slash, but testing this input is probably a good idea regardless. It would at least help to document this case.

(See @wthrowe's comment in header file)

File: tests/Unit/Utilities/Test_FileSystem.cpp
Comment (fmahebert) [inline tests/Unit/Utilities/Test_FileSystem.cpp]: Could also test `.hidden/path/to/dir/dummy.txt`, `.spectre`, `/test/path/to/dir/../dummy.txt`.

File: tests/Unit/Utilities/Test_FileSystem.cpp
Comment (fmahebert) [inline tests/Unit/Utilities/Test_FileSystem.cpp]: Could also test `dummy.txt` (no path), `.dummy.txt` (hidden), `./dummy.txt`, `../dummy.txt`

File: tests/Unit/Utilities/Test_FileSystem.cpp
Comment (fmahebert) [inline tests/Unit/Utilities/Test_FileSystem.cpp]: Filename advertises incorrect intent.

**Review by fmahebert — CHANGES_REQUESTED** [review]

File: tests/Unit/Utilities/Test_FileSystem.cpp
Comment (wthrowe) [reply tests/Unit/Utilities/Test_FileSystem.cpp]: Yeah, it looks like I forgot to make a note on it, but even after looking at the source it was still not clear to me what `get_path` does or is supposed to do.

File: src/Utilities/FileSystem.cpp
Comment (nilsdeppe) [reply src/Utilities/FileSystem.cpp]: To me it seems they should both test for `S_ISREG` and the documentation should state that this excludes symlinks, etc. and point to the POSIX documentation. Thoughts?

Comment (wthrowe) [reply src/Utilities/FileSystem.cpp]: `S_ISREG` seems reasonable, but saying that it excludes symlinks would be misleading, since `stat` resolves symlinks and will report a symlink to a regular file as a regular file (compare `lstat`).  Of course, a _broken_ symlink will be reported as a symlink.

Something like "is a regular file (or a link to one) (see stat(2))" would be my recommendation, preferably rewritten to look less like lisp.

File: src/Utilities/FileSystem.hpp
Comment (nilsdeppe) [reply src/Utilities/FileSystem.hpp]: Having thought about `get_path` and `get_file_name` I'm not sure they serve much purpose, definitely not `get_path` since it's not at all clear to me what `/path/to` should return. `get_file_name` suffers from the same ambiguity there but I think makes a bit more sense. Both can be handled by simple `string.find()` type manipulations. With that, I think we should just get rid of them because they don't really offer anything that is concrete.

File: tests/Unit/Utilities/Test_FileSystem.cpp
Comment (nilsdeppe) [reply tests/Unit/Utilities/Test_FileSystem.cpp]: See my reply to @wthrowe in header

File: tests/Unit/Utilities/Test_FileSystem.cpp
Comment (nilsdeppe) [reply tests/Unit/Utilities/Test_FileSystem.cpp]: See my reply to @wthrowe in header

File: src/Utilities/FileSystem.hpp
Comment (fmahebert) [reply src/Utilities/FileSystem.hpp]: The Python `os.path` module has various routines which may be interesting, if only as an example of functionality that someone considered useful: https://docs.python.org/2/library/os.path.html.

In particular, I think `os.path.split` may have a similar functionality to what you were trying to implement here. Combining the two functions into one reduces some of the ambiguity, but does not totally remove it.

I suspect there is some version of this function that is generically useful enough to be worth implementing (vs. every user writing their own `string.find()` manipulations explicitly), but I am not sure which version that would be.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: Use a vector so you don't have to do manual memory management.

File: src/Utilities/FileSystem.hpp
Comment (nilsdeppe) [reply src/Utilities/FileSystem.hpp]: Will and I talked about this in person and came up with a solution.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: Not all files have a dot in their names, and some directories do, so this might return "foo.bar/baz".  How about `^.*/([^/]+)`?

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: `free(name)`

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: Include `error` in the message.  Probably has to be saved since the `ERROR` internals might overwrite it.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: This should be a success case according to the documentation, since / presumably exists.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: free(the_cwd)

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: free(the_cwd)

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: Some thought on this function:
* Will we ever want failure to be nonfatal?  Even `rm` aborts on error.
* Most of the string manipulation appears to do the same thing as `dirname`.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: Is it safe to use `errno` inside `ERROR`?  I don't know what guarantees the standard library stuff gives about errno.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: What is `dir` for?  How is it different from `in_dir`?

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: While I think it is unlikely that we will want to create any directories directly in /, I don't see why we are just returning early if someone tries.

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: If we change the success case above to return then we will know getting here is an error and won't have to check.

File: src/Utilities/FileSystem.cpp
Comment (nilsdeppe) [reply src/Utilities/FileSystem.cpp]: I guess the only way `errno` could get overwritten is if `std::ostringstream` overwrites it when throwing an exception. I have no idea if this is something it is allowed to do or if it's even documented that it's allowed to do that. I'll add an extra variable just in case it can happen.

File: src/Utilities/FileSystem.cpp
Comment (nilsdeppe) [reply src/Utilities/FileSystem.cpp]: The goal was to be able to handle accidentally creating `/`. I've updated this, it now correctly handles `/` (or multiple: `////`) and also `/dir`

File: src/Utilities/FileSystem.cpp
Comment (wthrowe) [inline src/Utilities/FileSystem.cpp]: I still don't understand what this line is for.

Comment (nilsdeppe) [reply src/Utilities/FileSystem.cpp]: It's not. For some reason I thought it was because SpEC hand the while loop removing the slashes but that loop was only needed because the old code did all the parsing inline. Removed now and added a few more slashes randomly in the test path to make sure that extra slashes do not interfere with correctness.

**Comment by wthrowe** [comment]
Take the `dir` argument to create_directory by const ref and then this has my approval.

File: tests/Unit/Utilities/Test_FileSystem.cpp
Comment (kidder) [inline tests/Unit/Utilities/Test_FileSystem.cpp]: premissions -> permissions

File: src/Utilities/FileSystem.hpp
Comment (kidder) [inline src/Utilities/FileSystem.hpp]: why are you pointing to old documentation?
http://pubs.opengroup.org/onlinepubs/9699919799/functions/dirname.html

File: src/Utilities/FileSystem.hpp
Comment (kidder) [inline src/Utilities/FileSystem.hpp]: file_path -> path

File: src/Utilities/FileSystem.hpp
Comment (kidder) [inline src/Utilities/FileSystem.hpp]: doesn't it return the path to the parent directory of the parameter "path" (or "." if there is no '/' in "path"?

File: src/Utilities/FileSystem.hpp
Comment (kidder) [inline src/Utilities/FileSystem.hpp]: why not file_system?

File: src/Utilities/FileSystem.hpp
Comment (kidder) [inline src/Utilities/FileSystem.hpp]: I realize this wraps dirname, but wouldn't it be better to name it get_parent_path or something more descriptive?

File: src/Utilities/FileSystem.hpp
Comment (kidder) [inline src/Utilities/FileSystem.hpp]: what unit is wait_time in?

File: src/Utilities/FileSystem.hpp
Comment (kidder) [inline src/Utilities/FileSystem.hpp]: directory -> file

File: src/Utilities/FileSystem.hpp
Comment (kidder) [inline src/Utilities/FileSystem.hpp]: shouldn't the parameter be named path instead of file as we are determining whether it is a file?

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/Utilities/FileSystem.hpp
Comment (nilsdeppe) [reply src/Utilities/FileSystem.hpp]: I have no preference on the name of this function :) what you suggested is more likely to be understood by people not familiar with the POSIX function, while equally well understood by those who are. @wthrowe do you have an opinion?

Comment (wthrowe) [reply src/Utilities/FileSystem.hpp]: Something like `get_parent_path` seems more inline with our usual naming.  Let's do it.

Comment (nilsdeppe) [reply src/Utilities/FileSystem.hpp]: Alright, pushed updates. Once I have approval of the fixups I'll squash

**Review by kidder — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
pushed rebase

**Review by kidder — APPROVED** [review]

**Pull request merged**
